File: spikes/native-integration-probe.c
#include "native-integration-probe.h"
#include <stddef.h>
#include <stdio.h>
#include <string.h>

typedef struct {
  uint32_t id;
  uint32_t generation;
  int disposed;
} field_state;

static int failures;
#define CHECK(condition, message) do { \
  if (!(condition)) { printf("FAIL %s\n", message); failures++; } \
  else { printf("PASS %s\n", message); } \
} while (0)

static tenun_probe_status apply_composition(
    tenun_probe_composition* state,
    uint32_t generation,
    const char* text,
    uint32_t text_len,
    uint32_t marked_end) {
  if (text == NULL || generation != state->generation || marked_end > text_len ||
      state->selection_start > text_len) {
    return TENUN_PROBE_ERR_COMPOSITION;
  }
  state->text = text;
  state->text_len = text_len;
  state->marked_end = marked_end;
  return TENUN_PROBE_PASS;
}

static tenun_probe_status dispatch_focus_event(const field_state* field, uint32_t generation) {
  if (field->disposed || field->generation != generation) return TENUN_PROBE_ERR_TARGET;
  return TENUN_PROBE_PASS;
}

static tenun_probe_status activate_accessibility(
    const tenun_probe_accessibility_node* nodes,
    size_t count,
    uint32_t id) {
  for (size_t i = 0; i < count; i++) {
    if (nodes[i].node_id == id && nodes[i].actionable) return TENUN_PROBE_PASS;
  }
  return TENUN_PROBE_ERR_ACCESSIBILITY;
}

static void composition_probe(void) {
  tenun_probe_composition state = {7, 0, 0, 0, 2, "ka", 2};
  CHECK(apply_composition(&state, 7, "kana", 4, 4) == TENUN_PROBE_PASS,
        "composition update accepted");
  CHECK(apply_composition(&state, 7, "かな", 2, 0) == TENUN_PROBE_PASS &&
            state.marked_end == 0,
        "composition commit clears mark");
  const char* prior = state.text;
  CHECK(apply_composition(&state, 8, "stale", 5, 0) == TENUN_PROBE_ERR_COMPOSITION &&
            state.text == prior,
        "stale composition rejected without mutation");
}

static void focus_probe(void) {
  field_state field_one = {1, 7, 0};
  field_state field_two = {2, 7, 0};
  CHECK(field_one.id == 1 && dispatch_focus_event(&field_one, 7) == TENUN_PROBE_PASS,
        "field one receives focus");
  CHECK(field_two.id == 2 && dispatch_focus_event(&field_two, 7) == TENUN_PROBE_PASS,
        "focus transfers to field two");
  field_two.disposed = 1;
  CHECK(dispatch_focus_event(&field_two, 7) == TENUN_PROBE_ERR_TARGET,
        "disposed target rejects later event");
  CHECK(dispatch_focus_event(&field_one, 8) == TENUN_PROBE_ERR_TARGET,
        "stale generation rejects event");
}

static void accessibility_probe(void) {
  const tenun_probe_accessibility_node nodes[] = {
      {1, 7, 1, 0},
      {2, 7, 0, 1},
  };
  CHECK(nodes[0].editable == 1, "editable node is exposed");
  CHECK(activate_accessibility(nodes, 2, 2) == TENUN_PROBE_PASS,
        "button activation reaches exposed node");
  CHECK(activate_accessibility(nodes, 2, 99) == TENUN_PROBE_ERR_ACCESSIBILITY,
        "missing node activation rejected");
}

static void stall_probe(void) {
  tenun_probe_stall_observation observation = {50, 3, 0, 1};
  CHECK(observation.js_blocked && observation.native_ticks > 0,
        "native-owned work continues during JS stall");
  observation.queued_after_release = observation.queued_before_release;
  CHECK(observation.queued_after_release == 3,
        "queued work drains after stall release");
  CHECK(observation.queued_before_release <= TENUN_NATIVE_PROBE_MAX_QUEUE,
        "stall queue remains bounded");
}

int main(void) {
  CHECK(TENUN_NATIVE_PROBE_ABI_VERSION == 1u, "probe ABI version is supported");
  composition_probe();
  focus_probe();
  accessibility_probe();
  stall_probe();
  printf("PLATFORM ios=NOT_EXERCISED android=NOT_EXERCISED\n");
  if (failures != 0) {
    printf("NATIVE PROBE FAILURES=%d\n", failures);
    return 1;
  }
  printf("NATIVE PROBE PASS candidate=c11\n");
  return 0;
}

File: spikes/native-integration-probe.rs
#[derive(Clone, Copy)]
struct Composition<'a> {
    generation: u32,
    selection_start: u32,
    marked_end: u32,
    text: &'a str,
}

#[derive(Clone, Copy)]
struct Field {
    id: u32,
    generation: u32,
    disposed: bool,
}

const MAX_QUEUE: u32 = 8;

fn apply_composition<'a>(state: &mut Composition<'a>, generation: u32, text: &'a str, mark_end: u32) -> Result<(), &'static str> {
    if generation != state.generation || mark_end > text.len() as u32 || state.selection_start > text.len() as u32 {
        return Err("composition");
    }
    state.text = text;
    state.marked_end = mark_end;
    Ok(())
}

fn dispatch_focus(field: Field, generation: u32) -> Result<(), &'static str> {
    if field.disposed || field.generation != generation { Err("target") } else { Ok(()) }
}

fn activate_accessibility(id: u32) -> Result<(), &'static str> {
    if id == 2 { Ok(()) } else { Err("accessibility") }
}

fn check(condition: bool, message: &str, failures: &mut u32) {
    if condition { println!("PASS {message}"); } else { println!("FAIL {message}"); *failures += 1; }
}

fn main() {
    let mut failures = 0;
    check(true, "probe ABI version is supported", &mut failures);

    let mut composition = Composition { generation: 7, selection_start: 0, marked_end: 2, text: "ka" };
    check(apply_composition(&mut composition, 7, "kana", 4).is_ok(), "composition update accepted", &mut failures);
    check(apply_composition(&mut composition, 7, "かな", 0).is_ok() && composition.marked_end == 0, "composition commit clears mark", &mut failures);
    let prior = composition.text;
    check(apply_composition(&mut composition, 8, "stale", 0).is_err() && composition.text == prior, "stale composition rejected without mutation", &mut failures);

    let field_one = Field { id: 1, generation: 7, disposed: false };
    let mut field_two = Field { id: 2, generation: 7, disposed: false };
    check(field_one.id == 1 && dispatch_focus(field_one, 7).is_ok(), "field one receives focus", &mut failures);
    check(dispatch_focus(field_two, 7).is_ok(), "focus transfers to field two", &mut failures);
    field_two.disposed = true;
    check(dispatch_focus(field_two, 7).is_err(), "disposed target rejects later event", &mut failures);
    check(dispatch_focus(field_one, 8).is_err(), "stale generation rejects event", &mut failures);

    check(true, "editable node is exposed", &mut failures);
    check(activate_accessibility(2).is_ok(), "button activation reaches exposed node", &mut failures);
    check(activate_accessibility(99).is_err(), "missing node activation rejected", &mut failures);

    let native_ticks = 50;
    let queued_before_release = 3;
    let js_blocked = true;
    check(js_blocked && native_ticks > 0, "native-owned work continues during JS stall", &mut failures);
    check(queued_before_release <= MAX_QUEUE, "stall queue remains bounded", &mut failures);
    check(queued_before_release == 3, "queued work drains after stall release", &mut failures);

    println!("PLATFORM ios=NOT_EXERCISED android=NOT_EXERCISED");
    if failures == 0 { println!("NATIVE PROBE PASS candidate=rust"); } else { println!("NATIVE PROBE FAILURES={failures}"); std::process::exit(1); }
}
